contrib/babelfishpg_tsql/src/dbcmds.c

```diff
@@ -406,20 +386,61 @@ create_bbf_db_internal(const char *dbname, List *options, const char *owner, int
 	bool 			is_set_userid = false;
 	Oid 			save_userid;
 	const char	*old_createrole_self_grant;
+	ListCell	*option;
+	const char* database_collation_name = NULL;
 
-	/* TODO: Extract options */
-
-	tuple = SearchSysCache1(COLLOID, ObjectIdGetDatum(tsql_get_server_collation_oid_internal(false)));
-	if (!HeapTupleIsValid(tuple))
+	/* Check options */
+	foreach(option, options)
 	{
-		const char *server_collation_name = GetConfigOption("babelfishpg_tsql.server_collation_name", false, false);
+		DefElem    *defel = (DefElem *) lfirst(option);
 
-		ereport(ERROR,
-				(errcode(ERRCODE_UNDEFINED_OBJECT),
-				 errmsg("OID corresponding to collation \"%s\" does not exist", server_collation_name)));
+		if (strcmp(defel->defname, "collate") == 0)
+		{
+			database_collation_name = pstrdup(defGetString(defel));
+
+			if (tsql_find_collation_internal(database_collation_name) == NOT_FOUND)
+			{
+				database_collation_name = tsql_translate_tsql_collation_to_bbf_collation(database_collation_name);
+				if (tsql_find_collation_internal(database_collation_name) == NOT_FOUND)
+				{
+					ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("Invalid collation \"%s\"", defGetString(defel)),
+							parser_errposition(pstate, defel->location)));
+				}
+			}
+			if (!supported_collation_for_db_and_like(tsql_lookup_collation_table_internal(
+				get_collation_oid(list_make1(makeString((char*) database_collation_name)), false)).code_page))
+			{
+				const char *server_collation_name = GetConfigOption("babelfishpg_tsql.server_collation_name", false, false);
+				if (server_collation_name && strcmp(server_collation_name, database_collation_name))
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("\"%s\" is not currently supported for database collation ", defGetString(defel)),
+							parser_errposition(pstate, defel->location)));
+
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("\"%s\" is not currently supported for database collation ", defGetString(defel)),
+						parser_errposition(pstate, defel->location)));
+			}
+		}
+		else
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("option \"%s\" not recognized", defel->defname),
+					 parser_errposition(pstate, defel->location)));
+		}	
 	}
-	default_collation = ((Form_pg_collation) GETSTRUCT(tuple))->collname;
-	ReleaseSysCache(tuple);
+
+	if (database_collation_name == NULL)
+	{
+		database_collation_name = GetConfigOption("babelfishpg_tsql.server_collation_name", false, false);
+		if (tsql_find_collation_internal(database_collation_name) == NOT_FOUND)
+			database_collation_name = tsql_translate_tsql_collation_to_bbf_collation(database_collation_name);
```

Comment:
As explained above, we can simply this code. 

Additionally, if tsql_find_collation_internal could not find collation then something wrong must have happened as server_collation_name must exist. Raise error in that case.

Reply:
Updated